cpp/src/iterator/iterator.cuh

```diff
@@ -0,0 +1,443 @@
+    }
+};
+
+// @overload ColumnOutputMixed<T, false>
+template<typename T>
+struct ColumnOutputMixed<T, false>
+{
+    T value;
+    T value_squared;
+    gdf_index_type count;
+
+    template<typename T_input>
+    CUDA_HOST_DEVICE_CALLABLE
+    ColumnOutputMixed(T_input _value, bool is_valid)
+    : value( static_cast<T>(_value) )
+    {
+        value_squared = value*value;
+    };
+
+    CUDA_HOST_DEVICE_CALLABLE
+    ColumnOutputMixed(T _value, T _value_squared=0, gdf_index_type _count=0)
+    : value(_value), value_squared(_value_squared), count(_count)
+    {};
+
+    CUDA_HOST_DEVICE_CALLABLE
+    ColumnOutputMixed()
+    : value(0), value_squared(0), count(0)
+    {};
+
+    using this_t = ColumnOutputMixed<T, false>;
+
+    CUDA_HOST_DEVICE_CALLABLE
+    this_t operator+(this_t const &rhs) const
+    {
+        return this_t(
+            (this->value + rhs.value),
+            (this->value_squared + rhs.value_squared),
+            (this->count)
+        );
+    }
+
+    CUDA_HOST_DEVICE_CALLABLE
+    bool operator==(this_t const &rhs) const
+    {
+        return (
+            (this->value == rhs.value) &&
+            (this->value_squared == rhs.value_squared)
+        );
+    }
+};
+
+// --------------------------------------------------------------------------------------------------------
+/** -------------------------------------------------------------------------*
+ * @brief input struct with/without null bitmask
+ * A helper struct struct for `column_input_iterator`
+ * `ColumnInput.at(gdf_index_type id)` computes
+ * `data` value and valid flag at `id` and construct and return
+ * `T_output(data, valid flag)`.
+ * If nulls_present = true, the valid flag is always true.
+ * If nulls_present = false, the valid flag corresponds to null bitmask flag
+ * at `id`, and the data value = (is_valid(id))? data[id] : identity;
+ *
+ * @tparam  T_output a scalar data type to be output
+ * @tparam  T_element a native element type (cudf dta type) of input element array
+ *           and `identity` value which is used when null bitmaps flag is false.
+ * @tparam  nulls_present if true, this struct holds only data array.
+ *            else, this struct holds data array and bitmask array and identity value
+ * -------------------------------------------------------------------------**/
+template<typename T_output, typename T_element, bool nulls_present=true>
+struct ColumnInput;
+
+// @overload ColumnInput<T_output, T_element, false>
+template<typename T_output, typename T_element>
+struct ColumnInput<T_output, T_element, false>{
+    const T_element *data;
+
+    CUDA_HOST_DEVICE_CALLABLE
+    ColumnInput(const T_element *_data)
+    : data(_data){};
+
+    CUDA_HOST_DEVICE_CALLABLE
+    T_output at(gdf_index_type id) const {
+        return T_output(data[id], true);
+    };
+};
+
+// @overload ColumnInput<T_output, T_element, true>
+template<typename T_output, typename T_element>
+struct ColumnInput<T_output, T_element, true>{
```

Comment:
I don't think `ColumnInput` should have different member functions depending on `nulls_present`. That makes it difficult to use this struct uniformly whether or not it has nulls. `nulls_present` should simply be an optimization that avoids the `is_valid` check.

Reply:
Agree, will update.